rust/automerge/src/doctor.rs

```diff
@@ -0,0 +1,123 @@
+use itertools::Itertools;
+
+use crate::{
+    clocks::Clocks,
+    query,
+    types::{Clock, ObjId, Op, OpId, OpIds},
+    Automerge,
+};
+
+pub fn heal(doc: &mut Automerge) {
+    for mismatch in pred_mismatches(doc) {
+        let q = doc
+            .ops()
+            .search(&mismatch.obj, query::OpIdSearch::new(mismatch.op));
+        doc.ops_mut()
+            .change_vis(&mismatch.obj, q.index().unwrap(), |op| {
+                op.pred = mismatch.expected_preds.clone();
+            });
+    }
+}
+
+fn pred_mismatches(doc: &crate::Automerge) -> Vec<Mismatch> {
+    // An operation which sets a key which was previously set by a change in the depencies of the
+    // change containing the operation but which does not have a `pred` is suspicious. There's no
+    // way that such an operation would be generated by any well behaved library.
+    //
+    // scan through every op, building a list of operations which _should_ precede each op
+    let clocks = doc.build_clocks();
+
+    let mut keyops = Vec::new();
+    let mut mismatches = Vec::new();
+    for (obj, _typ, ops) in doc.ops().iter_objs() {
+        let ops_by_key = ops.group_by(|o| o.key);
```

Comment:
This will work for maps but not for lists - should be `group_by(|o| o.elemid_or_key())` since the first op in a list that has the insert flag has a different key than the rest and could get grouped with the previous elemid